client/src/lib.rs

```diff
@@ -69,31 +91,50 @@ impl<C: Clone + Deref<Target = impl Signer>> Client<C> {
         }
     }
 
-    pub fn program(&self, program_id: Pubkey) -> Program<C> {
-        Program {
-            program_id,
-            cfg: Config {
-                cluster: self.cfg.cluster.clone(),
-                options: self.cfg.options,
-                payer: self.cfg.payer.clone(),
-            },
-        }
+    pub fn program(&self, program_id: Pubkey) -> Result<Program<C>, ClientError> {
+        let cfg = Config {
+            cluster: self.cfg.cluster.clone(),
+            options: self.cfg.options,
+            payer: self.cfg.payer.clone(),
+        };
+
+        Program::new(program_id, cfg)
     }
 }
 
 // Internal configuration for a client.
 #[derive(Debug)]
-struct Config<C> {
+pub struct Config<C> {
     cluster: Cluster,
     payer: C,
     options: Option<CommitmentConfig>,
 }
 
+pub struct EventUnsubscriber<'a> {
+    handle: JoinHandle<Result<(), ClientError>>,
+    rx: UnboundedReceiver<UnsubscribeFn>,
+    #[cfg(not(feature = "async"))]
+    runtime_handle: &'a Handle,
+    _lifetime_marker: PhantomData<&'a Handle>,
+}
+
+impl<'a> EventUnsubscriber<'a> {
+    async fn unsubscribe_internal(mut self) {
+        if let Some(unsubscribe) = self.rx.recv().await {
+            unsubscribe().await;
+        }
+
+        let _ = self.handle.await;
+    }
+}
+
 /// Program is the primary client handle to be used to build and send requests.
-#[derive(Debug)]
 pub struct Program<C> {
     program_id: Pubkey,
     cfg: Config<C>,
+    sub_client: Mutex<Option<Arc<PubsubClient>>>,
```

Comment:
`Mutex<Option<Arc<PubsubClient>>>` is a bit off I think.

- Because we set the value once and only read afterwards, I think using `RwLock` is much better for this case as it allows multiple readers at the same time(and is optimized for it) meanwhile `Mutex` only allows a single reader.
- We'd like to hold a reference to the entire value so it's better to have the `Arc` outside. 

So the final `sub_client` would be:

```suggestion
    sub_client: Arc<RwLock<Option<PubsubClient>>>,
```